src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,408 @@
+        }
+    }
+
+    /**
+     * Reads the inputs passed from a file and adds the tasks into the TaskArray.
+     *
+     * @param s The input line.
+     */
+    public void parseFromFile(String s) {
+        String[] chars = s.split(" / ");
+        String type = chars[0];
+        boolean isDone = chars[1].equals("1");
+        DateTimeFormatter dateFormat = DateTimeFormatter.ofPattern("HH:mm");
+        Task t;
+        switch (type) {
+        case "[T]":
+            t = new Todo(chars[2]);
+            if (isDone) {
+                t.markDone();
+            }
+            tasks.add(t);
+            break;
+        case "[D]":
+            if (chars.length == 4) {
+                t = new Deadline(chars[2], LocalDate.parse(chars[3]));
+            } else {
+                t = new Deadline(chars[2], LocalDate.parse(chars[3]),
+                        LocalTime.parse(chars[4], dateFormat));
+            }
+            if (isDone) {
+                t.markDone();
+            }
+            tasks.add(t);
+            break;
+        case "[E]":
+            if (chars.length == 7) {
+                t = new Event(chars[2], LocalDate.parse(chars[3]), LocalTime.parse(chars[4], dateFormat),
+                        LocalDate.parse(chars[5]), LocalTime.parse(chars[6]));
+            } else if (chars.length == 5) {
+                t = new Event(chars[2], LocalDate.parse(chars[3]), LocalDate.parse(chars[4]));
+            } else {
+                if (chars[5].length() > 5) {
+                    t = new Event(chars[2], LocalDate.parse(chars[3]), LocalTime.parse(chars[4], dateFormat), //chars
+                            LocalDate.parse(chars[5]));
+                } else {
+                    t = new Event(chars[2], LocalDate.parse(chars[3]), LocalDate.parse(chars[4]),
+                            LocalTime.parse(chars[5], dateFormat));
+                }
+            }
+            if (isDone) {
+                t.markDone();
+            }
+            tasks.add(t);
+            break;
+        }
+    }
+
+    /**
+     * Prints out the tasks in the TaskArray.
+     */
+    private void printList() {
+        if (tasks.isEmpty()) {
+            this.ui.print("list is empty :(");
+        } else {
+            for (int i = 0; i < tasks.size(); i++) {
+                this.ui.print(i + 1 + " " + tasks.get(i).toString());
+            }
+        }
+    }
+
+    private void ending() {
+        this.isExit = true;
+        this.ui.farewell();
+    }
+
+    /**
+     * Adds a todo task into the list.
+     *
+     * @param x Details of the task.
+     */
+    private void addTodo(String x) {
+        if (x == null || x.trim().isEmpty()) {
+            throw new LackDescriptionException("todo");
+        }
+        Todo t = new Todo(x);
+        tasks.add(t);
+        addedTask(x);
+    }
+
+    /**
+     * Adds a deadline task into the list.
+     *
+     * @param x Details of the task.
+     */
+    private void addDeadline(String x) {
```

Comment:
Maybe can improve the name of the parameter so that its more clear and easy to follow? Same comment for other functions.